### ISSUE TITLE
Possible resource limitation for very large import files

### ISSUE THREAD

**Issue by Docmale** [digitalfabrik/entitlementcard#1610]
**Describe the bug**
Error: 
![image (2)](https://github.com/user-attachments/assets/d27b89d6-9eb1-4e87-8f91-c9aaeacdf74d)
I got the error with two files with 299 and 300 entries. When I tried again with 150 entries and later with 300 entries, I could no longer reproduce the behaviour. The error only occurs again when I start another generation in parallel. So it has something to do with resources?

**Comment by Docmale** [comment]
The effect occurs with Freinet and standard CSV import

**Comment by ztefanie** [comment]
Does not occur every time, just sometimes.

**Comment by f1sh1918** [comment]
created cards from csv (300 entries) in two browser tabs in parallel. Couldn't reproduce the error with chrome

**Comment by seluianova** [comment]
Couldn't reproduce locally, but the issue is reproduced on staging.
The error occurs because of the timeout (it seems to be 1 minute).
`POST https://api.staging.entitlementcard.app/ 504 (Gateway Timeout)`
I guess it's limited on the infrastructure level (?)
@f1sh1918 do you know if we can (and want to) adjust it?

**Comment by f1sh1918** [comment]
> Couldn't reproduce locally, but the issue is reproduced on staging.
> The error occurs because of the timeout (it seems to be 1 minute).
> `POST https://api.staging.entitlementcard.app/ 504 (Gateway Timeout)`
> I guess it's limited on the infrastructure level (?)
> @f1sh1918 do you know if we can (and want to) adjust it?

Can you please ask that in our tech channel. I do not know that.

**Comment by seluianova** [comment]
> This can also be an issue for larger csv files maybe? Not sure

Indirectly yes. The error occurs when it takes more than 1 minute to process a request. Which is most likely to happen when the file is large.

**Comment by seluianova** [comment]
> This can also be an issue for larger csv files maybe? Not sure

@f1sh1918 it could be, but we don't allow to upload more than 300 entries in one file

**Comment by seluianova** [comment]
Another analysis round:
There seems to be a significant performance degradation on the staging env.
Processing of a single request with 300 entries takes ~40 seconds locally:
![Image](https://github.com/user-attachments/assets/a25b71b2-d0a4-45d6-93fa-78f55e553579)
While the exact same request on staging takes ~60 seconds:
![Image](https://github.com/user-attachments/assets/6bbcf938-5d86-4554-b731-fc5c24bc75e9)

**Comment by seluianova** [comment]
As discussed with Sven, the longer response time on staging env is related to CPU limitation on the test server.

I see the following ways to resolve the current task:

1. leave as is, because most probably this issue won't be reproduced on production (we have more CPU resources there)
2. rescale test server --> probably not needed, if we don't get any complains from users

I also did some profiling and found that 97% of the processing time is spent hashing the activation code.
```
    fun hashActivationSecret(rawActivationSecret: ByteArray): ByteArray =
        BCrypt.withDefaults().hash(cost, rawActivationSecret)
```
We currently use a hash cost of 11, and I'm not sure this is really necessary in our case, given that the value we're hashing is generated randomly on our side.
The default cost is 10, and if we use it, it cuts the request processing time in half.

3. so, my third suggestion would be: set the hashing cost to 10, and that will also solve the current problem.

**Comment by f1sh1918** [comment]
Wow great investigation. 👍 

I think for passwords 11 is (was) a good choice.
Since we hash an object with lots of complexity and a secure random, 10 may be good enough though.
But on the other hand we are hashing sensitive data.
Since we have no "real" performance issues, i dunno if we should really touch that.

But maybe @michael-markl , @svenseeberg or @maxammann can bring more light into that :)

Here i found a nice article about that topic. 
https://labs.clio.com/bcrypt-cost-factor-4ca0a9b03966

Due to increasing cpu performance a recommendation was to increase the cost factor based on the todays date and f.e. increase it every 18 month.

**Comment by michael-markl** [comment]
Yeah, I'm not sure if BCrypt was necessary in the first place :D With 20 bytes of randomness, it might have been enough to just use SHA256, but no idea :D

For this problem, I think it's enough to simply extend the timeout period to several minutes (probably an nginx config?)

**Comment by seluianova** [comment]
> But on the other hand we are hashing sensitive data like name and birthday.

@f1sh1918 in the current case we are hashing a randomly generated byte array.
```
        val secureRandom = SecureRandom.getInstanceStrong()
        val pepper = ByteArray(PEPPER_LENGTH)
        secureRandom.nextBytes(pepper)

        val rawActivationSecret = ByteArray(activationSecretLength)
        secureRandom.nextBytes(rawActivationSecret)

        val activationSecretHash = CardActivator.hashActivationSecret(rawActivationSecret)
```

**Comment by maxammann** [comment]
I'd need to look closer into why we went with bcrypt instead of sha256. Moving away from this would be a breaking change though.

I think the proper fix would be to asynchronously process the job and poll for it's completion.

Changing rounds with bcrypt should be backwards compatible

**Comment by michael-markl** [comment]
> I'd need to look closer into why we went with bcrypt instead of sha256. Moving away from this would be a breaking change though.

There was no particular discussion about which hashing method we use (see https://github.com/digitalfabrik/entitlementcard/pull/804 and related issue); a salt is definitely good, so I assume bcrypt is definitely better than a "raw" sha256.

> I think the proper fix would be to asynchronously process the job and poll for it's completion.

Maybe, but it's also a quite the complexity increase :D The job "result" contains many secrets, so it needs to / should be cleaned up if the user doesn't fetch the result in some meaningful time... also, we probably don't want to persist the secrets, so if the backend restarts due to a crash/maintenance, the (intermediate) results are gone?

> Changing rounds with bcrypt should be backwards compatible

Would be my favorite option so far (unless the crypto turns too bad of course)

**Comment by seluianova** [comment]
> I could test that tomorrow

I have created 2 cards - with a cost of 10 and 11.
Both can be activated and successfully verified.
